src/common/hist_util.cc

```diff
@@ -275,10 +321,18 @@ void DenseCuts::Build(DMatrix* p_fmat, uint32_t max_num_bins) {
             // move to next group
             group_ind++;
           }
-          for (auto const& entry : inst) {
-            if (entry.index >= begin && entry.index < end) {
-              size_t w_idx = use_group ? group_ind : ridx;
-              sketchs[entry.index].Push(entry.fvalue, info.GetWeight(w_idx));
+          size_t w_idx = use_group ? group_ind : ridx;
+          auto w = info.GetWeight(w_idx);
+          if (isDense) {
+            auto data = inst.data();
+            for (size_t ii = begin; ii < end; ii++) {
+              sketchs[ii].Push(data[ii].fvalue, w);
+            }
+          } else {
+            for (auto const& entry : inst) {
+              if (entry.index >= begin && entry.index < end) {
+                sketchs[entry.index].Push(entry.fvalue, w);
+              }
```

Comment:
Why is this necessary?  I thought it has nothing to do with column matrix?

Reply:
I have noticed that for dense matrix we should not iterate through over all feature for all thread to get right `entry.fvalue`.
Before this fix we had complexity O(n_features)  for each thread (all elements should be loaded and compared with `begin`, `end`), but now O(nstep), where nstep ~ n_feature/nthreads.
It did a performance gain of course, sorry but I don't remember exactly how much, but if it's critical I could check it in case reduced complexity is not enough.